### PR TITLE
UCP/API: Add option to specify expected number of endpoints.

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

File: src/ucp/api/ucp.h [hunk 1]
```diff
@@ -470,6 +471,14 @@ typedef struct ucp_params {
      * then this context needs thread safety.
      */
     int                                mt_workers_shared;
+
+    /**
+     * An optimization hint of how many endpoints would be created on this context.
+     * Does not affect semantics, but only transport selection criteria and the
+     * resulting performance.
```
Comment (shamisp) [inline src/ucp/api/ucp.h]: I would elaborate a bit. You can mention that the number can be extracted from mpirun -np ?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3034/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/577/ for details.

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2518/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/578/ for details.

**Comment by yosefe** [comment]
bot:ornl:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr/3040/ for details.

**Comment by mellanox-github** [comment]
Test FAILed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/585/ for details.

**Comment by yosefe** [comment]
bot:ornl:retest

**Comment by yosefe** [comment]
bot:bgateNEW:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/gh-ucx-pr__NEW/592/ for details.

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2532/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2536/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2545/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2549/ for details (Mellanox internal link).

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by yosefe** [comment]
bot:mlx:retest

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2559/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2561/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/2563/ for details (Mellanox internal link).

**Comment by brminich** [comment]
:+1:

**Comment by shamisp** [comment]
👍

**Comment by shamisp** [comment]
API Change has to be approved by @bbenton

**Comment by bbenton** [comment]
:+1:

**Pull request merged**